drivers/SmartThings/philips-hue/src/disco/button.lua

```diff
@@ -0,0 +1,148 @@
+local log = require "logjam"
+local socket = require "cosock".socket
+local st_utils = require "st.utils"
+
+local HueDeviceTypes = require "hue_device_types"
+
+---@class DiscoveredButtonHandler: DiscoveredChildDeviceHandler
+local M = {}
+
+---@param api_instance PhilipsHueApi
+---@param device_service_info HueDeviceInfo
+---@param bridge_id string
+---@param resource_id string
+---@param cache table?
+---@return table<string,any>? description nil on error
+---@return string? err nil on success
+local function _do_update(api_instance, device_service_info, bridge_id, resource_id, cache)
+  local rid_by_rtype = {}
+  local button_services = {}
+  local num_buttons = 0
+
+  for _, svc in ipairs(device_service_info.services) do
+    if svc.rtype == HueDeviceTypes.BUTTON then
+      num_buttons = num_buttons + 1
+      table.insert(button_services, svc.rid)
+    else
+      rid_by_rtype[svc.rtype] = svc.rid
+    end
+  end
+
+  local button_remote_description = {
+    hue_provided_name = device_service_info.metadata.name,
+    parent_device_id = bridge_id,
+    hue_device_id = device_service_info.id,
+    hue_device_data = device_service_info,
+    id = resource_id,
+    num_buttons = num_buttons
+  }
+
+  for _, button_rid in ipairs(button_services) do
+    local button_repr, err = api_instance:get_button_by_id(button_rid)
+    if err or not button_repr then
+      log.error("Error getting button representation: " .. tostring(err or "unknown error"))
+    else
+      local control_id = button_repr.data[1].metadata.control_id
+      local button_key = string.format("button%s", control_id)
+      local button_id_key = string.format("%s_id", button_key)
+      button_remote_description[button_key] = button_repr.data[1].button
+      button_remote_description[button_id_key] = button_repr.data[1].id
+
+      if control_id == 1 and button_remote_description.id == nil then
+        button_remote_description.id = button_repr.data[1].id
+      end
+    end
+  end
+
+  local battery, battery_err = api_instance:get_device_power_by_id(rid_by_rtype[HueDeviceTypes.DEVICE_POWER])
+  if battery_err then return nil, battery_err end
```

Comment:
Does the caching below need to happen regardless of this?  Or is it always expected that there be device power info and this is a fatal error?

Reply:
An error here would indicate a connectivity error, not a REST/HTTP error. So there would be no point in continuing becuase we would expect additional REST calls to fail, since the REST client itself has retry heuristics built in.

Reply:
I confirmed that as long as a `GET` REST call returns JSON, you shouldn't get an error here; you'll get a valid response with errors in the JSON payload and no `data` array.

- All `GET` REST calls converge on `do_get`, which sends the HTTP request to a separate task over a channel: https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/038217a0da0b337e3b3d7b569fea847d73a7d142/drivers/SmartThings/philips-hue/src/hue/api.lua#L207-L222
- The REST call has retry logic, and then tries to forward *whatever* it gets back at the end of that to `process_rest_response`: https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/038217a0da0b337e3b3d7b569fea847d73a7d142/drivers/SmartThings/philips-hue/src/hue/api.lua#L167-L171
- As long as the response has a body and that body is valid JSON, it will not return an error: https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/038217a0da0b337e3b3d7b569fea847d73a7d142/drivers/SmartThings/philips-hue/src/hue/api.lua#L102-L116
- Querying the Hue Bridge for a non-existent resource does give an HTTP 404, but we're not currently looking at that for error conditions; and the 404 includes a valid JSON body: 
<img width="1271" alt="image" src="https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/assets/444635/c4b93bd0-aeb9-4472-993c-df43b02279ea">